prow/git/publisher.go

```diff
@@ -0,0 +1,154 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package git
+
+import (
+	"fmt"
+	"net/url"
+)
+
+// RemoteResolver knows how to construct a remote URL for git calls
+type RemoteResolver interface {
+	Resolve() (string, error)
+}
+
+// StaticURLFor returns a generator that builds a static URL from the inputs.
+// This is useful when remote URLs never change, like when referring to an
+// authoritative clone URL for a repo.
+func StaticURLFor(host, org, repo string) func() (*url.URL, error) {
+	return func() (*url.URL, error) {
+		return &url.URL{Host: host, Path: fmt.Sprintf("%s/%s", org, repo)}, nil
+	}
+}
+
+// DynamicURLFor returns a generator that builds a dynamic URL from the inputs.
+// This is useful when remote URLs change, like when referring to a user's fork
+// of a repository.
+func DynamicURLFor(host, repo string, org func() (string, error)) func() (*url.URL, error) {
+	return func() (*url.URL, error) {
+		o, err := org()
+		if err != nil {
+			return nil, err
+		}
+		return &url.URL{Host: host, Path: fmt.Sprintf("%s/%s", o, repo)}, nil
+	}
+}
+
+// NewSimpleAuthResolver builds a dynamic resolver that knows how to
+// create simple auth URLs
+func NewSimpleAuthResolver(remote func() (*url.URL, error), username func() (string, error), token func() []byte) RemoteResolver {
+	return &simpleAuthResolver{
+		remote:   remote,
+		username: username,
+		token:    token,
+	}
+}
+
+// simpleAuthResolver builds URLs with simple auth credentials, resolved
+// dynamically. This structure is threadsafe.
+type simpleAuthResolver struct {
+	remote   func() (*url.URL, error)
+	username func() (string, error)
+	token    func() []byte
+}
+
+func (r *simpleAuthResolver) Resolve() (string, error) {
+	// copy the base endpoint to ensure we don't mutate it
+	remote, err := r.remote()
+	if err != nil {
+		return "", fmt.Errorf("could not resolve remote: %v", err)
+	}
+	name, err := r.username()
+	if err != nil {
+		return "", fmt.Errorf("could not resolve username: %v", err)
+	}
+	remote.User = url.UserPassword(name, string(r.token()))
+	return remote.String(), nil
+}
+
+// NewSSHResolver builds a resolver for SSH remotes
+func NewSSHResolver(host, repo string, org func() (string, error)) RemoteResolver {
+	return &sshResolver{
+		host: host,
+		repo: repo,
+		org:  org,
+	}
+}
+
+// sshResolver holds static URLs for SSH interaction
+type sshResolver struct {
+	host, repo string
+	org        func() (string, error)
+}
+
+func (r *sshResolver) Resolve() (string, error) {
+	org, err := r.org()
+	if err != nil {
+		return "", err
+	}
+	return fmt.Sprintf("git@%s:%s/%s.git", r.host, org, r.repo), nil
+}
+
+// fakeResolver allows for simple injections in tests
+type fakeResolver struct {
```

Comment:
Nit: Maybe put this in the `_test` file?

Reply:
I imagine this will be used in a multitude of test files as this implementation style takes over the other one, so I wanted it here for ease of discovery.